here-naksha-app-service/src/test/java/com/here/naksha/app/common/ResponseAssertions.java

```diff
@@ -71,4 +83,97 @@ public ResponseAssertions hasJsonBody(String expectedJsonBody, String failureMes
+      if (prefixId != null) {
+        assertTrue(
+                insertedIds.get(i).startsWith(prefixId),
+                "Inserted Feature Id in the response doesn't start with given prefix Id : " + prefixId);
+      }
+      assertEquals(
+              insertedIds.get(i),
+              features.get(i).getId(),
+              "Mismatch between inserted v/s feature ID in the response at idx : " + i);
+    }
+    return this;
+  }
+
+  public ResponseAssertions hasMatchingUpdatedCount(int cnt) throws JSONException {
+    JSONAssert.assertEquals("{updated:[" + cnt + "]}", subject.body(),
+            new ArraySizeComparator(JSONCompareMode.LENIENT));
+    return this;
+  }
+
+  public ResponseAssertions hasUpdatedIdsMatchingFeatureIds(final @Nullable String prefixId) {
+    if (collectionResponse==null) collectionResponse = parseJson(subject.body(), XyzFeatureCollection.class);
+    final List<String> updatedIds = collectionResponse.getUpdated();
+    final List<XyzFeature> features = collectionResponse.getFeatures();
+    for (int i = 0; i < updatedIds.size(); i++) {
+      if (prefixId != null) {
+        assertTrue(
+                updatedIds.get(i).startsWith(prefixId),
+                "Updated Feature Id in the response doesn't start with given prefix Id : " + prefixId);
+      }
+      assertEquals(
+              updatedIds.get(i),
+              features.get(i).getId(),
+              "Mismatch between updated v/s feature ID in the response at idx : " + i);
+    }
+    return this;
+  }
+
+  public ResponseAssertions hasFeatureReferencedByViolations(int featureIdx, int[] violationIndices) {
+    if (collectionResponse==null) collectionResponse = parseJson(subject.body(), XyzFeatureCollection.class);
+    final List<XyzFeature> features = collectionResponse.getFeatures();
+    final List<XyzFeature> violations = collectionResponse.getViolations();
+    // obtain feature Id
+    final String fId = features.get(featureIdx).getId();
+    assertNotNull(fId, "Feature Id at index "+featureIdx+" found null");
+    // match feature Id with references of Violation object at given index positions
+    assertNotNull(violations, "No violations found in response");
+    for (int i = 0; i < violationIndices.length; i++) {
+      int vIdx = violationIndices[i];
+      final List<XyzReference> references = violations.get(vIdx).getProperties().getReferences();
+      assertNotNull(references, "References missing for violation at idx " + vIdx);
+      for (final XyzReference reference : references) {
+        assertNotNull(reference.getId(), "Id missing in references for violation at idx " + vIdx);
+        assertEquals(fId, reference.getId(), "Referenced featured id doesn't match for Violation at idx " + vIdx);
+      }
+    }
+    return this;
+  }
+
+
+  public ResponseAssertions hasUuids() {
+    if (collectionResponse==null) collectionResponse = parseJson(subject.body(), XyzFeatureCollection.class);
```

Comment:
nth: moving collectionResponse init into the private constructor might be good, since it is needed in many assertions

Reply:
Actually, I think, we don't even need a constructor for `collectionResponse`. If you see at class level, this is cache'd once parsed, so next function call will not require it to be parsed again. And this is better than someone having to pass it explicitly as constructor (which already accepts entire response body).

Reply:
I mean straight up parsing and caching the `collectionResponse` might be simpler

```
  private ResponseAssertions(HttpResponse<String> subject) {
    this.subject = subject;
    this.collectionResponse = parseJson(subject.body(), XyzFeatureCollection.class);
  }
```

But yes nothing major, the code is fine as is :)

Reply:
It is not done this way, as not all responses have FeatureCollection.